### PR TITLE
feat(upgrade): add rebuild and single volume replica checks

### PR DESCRIPTION
**Pull request by sinhaashish**
Screenshot 2023-02-22 at 5 17 03 PM

### CODE REVIEW

File: k8s/plugin/src/upgrade_validations.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+pub async fn upgrade_validations() -> Result<UpgradeValidation, Error> {
+    let single_replica_volumes = Vec::new();
+    let mut rebuild_in_progress_in_cluster = false;
+    let mut validation = UpgradeValidation {
+        rebuild_in_progress_in_cluster,
+        single_replica_volumes,
+    };
+    match RestClient::client()
+        .volumes_api()
+        .get_volumes(0, None, None)
```
Comment (tiagolobocastro) [inline k8s/plugin/src/upgrade_validations.rs:33]: Should we use paging here?

File: k8s/plugin/src/upgrade_validations.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        .await
+    {
+        Ok(volumes) => {
+            for volume in volumes.into_body().entries {
+                if volume.state.target.is_some() {
+                    let target = volume.state.target.as_ref().unwrap();
```
Comment (tiagolobocastro) [inline k8s/plugin/src/upgrade_validations.rs:39]:
```suggestion
                if let Some(target) = &volume.state.target {
```

File: k8s/plugin/src/upgrade_validations.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    pub rebuild_in_progress_in_cluster: bool,
+    pub single_replica_volumes: Vec<Uuid>,
+}
+
+/// Cluster lever validation before upgrade.
+/// 1. Check for any volume rebuild in progress across the cluster
+/// 2. List all single replica volumes
+pub async fn upgrade_validations() -> Result<UpgradeValidation, Error> {
+    let single_replica_volumes = Vec::new();
+    let mut rebuild_in_progress_in_cluster = false;
```
Comment (tiagolobocastro) [inline k8s/plugin/src/upgrade_validations.rs:26]: move variable declaration close to their actual usage.

File: k8s/plugin/src/upgrade_validations.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        .await
+    {
+        Ok(volumes) => {
+            for volume in volumes.into_body().entries {
+                if volume.state.target.is_some() {
+                    let target = volume.state.target.as_ref().unwrap();
+                    for child in &target.children {
+                        rebuild_in_progress_in_cluster =
+                            rebuild_in_progress_in_cluster || child.rebuild_progress.is_some();
+                    }
```
Comment (tiagolobocastro) [inline k8s/plugin/src/upgrade_validations.rs:43]:
```suggestion
      validation.rebuild_in_progress = target.children.iter().any(|child| child.rebuild_progress.is_some());
```

File: k8s/plugin/src/main.rs [hunk 5]
```diff
@@ -167,7 +168,32 @@ async fn execute(cli_args: CliArgs) {
                 }
             },
             Operations::Upgrade(resources) => {
-                console_logger::info(user_prompt::UPGRADE_WARNING);
+                console_logger::info(upgrade_validations::UPGRADE_WARNING);
+
+                match upgrade_validations::upgrade_validations().await {
+                    Ok(upgrade_validations) => {
+                        if upgrade_validations.rebuild_in_progress_in_cluster {
```
Comment (tiagolobocastro) [inline k8s/plugin/src/main.rs:175]: nit: name is quite verbose, suggest reducing, eg: rebuild_in_progress

**Review by tiagolobocastro — COMMENTED** [review]
Plugin is getting quite chunky, we need to remove operator logic into its own crate.
Someone should raise a PR for that. Any takers?

**Comment by sinhaashish** [comment]
> Plugin is getting quite chunky, we need to remove operator logic into its own crate. Someone should raise a PR for that. Any takers?

Taken care in PR https://github.com/openebs/mayastor-extensions/pull/117 . Its merged

File: k8s/plugin/src/upgrade_validations.rs (hunk 1 above)
Comment (sinhaashish) [reply k8s/plugin/src/upgrade_validations.rs]: I knew only this api . Is there any other api for getting the volumes ?

File: k8s/upgrade/src/upgrade_validations.rs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+/// Warning to users before doing an upgrade
+pub const SINGLE_REPLICA_VOLUME_WARNING: &str =  "The list below shows the single replica volumes in cluster.\nThese single replica volumes may not be accessible during upgrade.\nPlease retry  with '--skip-single-replica-volume-validation` flag to skip this validation.\n";
+
+#[derive(Debug)]
+pub struct UpgradeValidation {
+    pub rebuild_in_progress: bool,
+    pub single_replica_volumes: Vec<Uuid>,
+}
+
+/// Cluster lever validation before upgrade.
```
Comment (Abhinandan-Purkait) [inline k8s/upgrade/src/upgrade_validations.rs:21]: lever ---> level

File: k8s/upgrade/src/upgrade_validations.rs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        Ok(volumes) => {
+            for volume in volumes.into_body().entries {
+                if let Some(target) = &volume.state.target {
+                    validation.rebuild_in_progress = target
+                        .children
+                        .iter()
+                        .any(|child| child.rebuild_progress.is_some());
+                }
+
+                if volume.spec.num_replicas == SINGLE_REPLICA_VOLUME {
```
Comment (Abhinandan-Purkait) [inline k8s/upgrade/src/upgrade_validations.rs:44]: Does degraded volumes with one replica left also not qualify as Single Replica Volume during upgrade? cc @tiagolobocastro  @avishnu

Comment (avishnu) [reply k8s/upgrade/src/upgrade_validations.rs:44]: I think there are 2 aspects to this:
1. If rebuild is in-progress, it is handled through a different check.
2. If rebuild is not yet started or cannot happen, then it needs to be flagged.

**Pull request closed without merging**